File: src/GKSheetList.pas
unit GKSheetList;

interface

uses
  Types, Classes, Graphics, Controls, ComCtrls, ActnList, Menus, StdCtrls,
  bsCtrls, GKCommon;

type
  TModifyEvent = procedure (Sender: TObject; ItemData: TObject; Action: TRecAction) of object;

  TListButtons = set of (lbAdd, lbEdit, lbDelete, lbJump, lbMoveUp, lbMoveDown);

  TListMode = (lmView, lmBox);

  TSheetList = class(TCustomControl)
  private
    //FActionList: TActionList;
    FActionAdd: TAction;
    FActionDelete: TAction;
    FActionEdit: TAction;
    FActionJump: TAction;
    FActionMoveUp: TAction;
    FActionMoveDown: TAction;

    FBtnAdd: TToolButton;
    FBtnDelete: TToolButton;
    FBtnEdit: TToolButton;
    FBtnLinkJump: TToolButton;
    FBtnMoveUp: TToolButton;
    FBtnMoveDown: TToolButton;

    FList: TCustomListControl;
    FToolBar: TToolBar;

    FOnModify: TModifyEvent;
    FButtons: TListButtons;

    FListMode: TListMode;
    FReadOnly: Boolean;

    procedure ButtonClick(Sender: TObject);
    procedure ListDblClick(Sender: TObject);
    procedure ListKeyDown(Sender: TObject; var Key: Word; Shift: TShiftState);
    //procedure SetActionList(const Value: TActionList);
    procedure SheetShow(Sender: TObject);
    procedure SetButtons(const Value: TListButtons);

    function GetSelectedData(): TObject;

    procedure LBDrawItem(Control: TWinControl; Index: Integer; Rect: TRect;
      State: TOwnerDrawState);
    procedure LBMeasureItem(Control: TWinControl; Index: Integer;
      var Height: Integer);
    procedure SetReadOnly(const Value: Boolean);
  protected
  public
    constructor Create(AOwner: TComponent; aListMode: TListMode = lmView);
    destructor Destroy; override;

    //property ActionList: TActionList read FActionList write SetActionList;
    property Buttons: TListButtons read FButtons write SetButtons;
    property List: TCustomListControl read FList;
    property ToolBar: TToolBar read FToolBar;

    procedure ItemAdd();
    procedure ItemEdit();
    procedure ItemDelete();
    procedure ItemJump();
    procedure ItemMoveUp();
    procedure ItemMoveDown();

    procedure Columns_BeginUpdate();
    procedure Columns_Clear();
    procedure Columns_EndUpdate();

    procedure AddColumn(aCaption: string; aWidth: Integer; aAutoSize: Boolean = False);

    property OnModify: TModifyEvent read FOnModify write FOnModify;
    property ReadOnly: Boolean read FReadOnly write SetReadOnly; 
  end;

procedure AddListColumn(aList: TCustomListControl; aCaption: string; aWidth: Integer;
  aAutoSize: Boolean = False);

procedure ResizeColumn(aList: TBSListView; aColumnIndex: Integer);
  
implementation

uses
  Windows, GKMain, ExtCtrls;

procedure AddListColumn(aList: TCustomListControl; aCaption: string; aWidth: Integer;
  aAutoSize: Boolean = False);
begin
  if not(aList is TBSListView) then Exit;

  with TBSListView(aList).Columns.Add() do begin
    Caption := aCaption;
    Width := aWidth;
    AutoSize := aAutoSize;
  end;
end;

procedure ResizeColumn(aList: TBSListView; aColumnIndex: Integer);
var
  i, max_w, w: Integer;
  item: TListItem;
begin
  if (aColumnIndex < 0) or (aColumnIndex >= aList.Columns.Count) then Exit;

  max_w := 0;

  for i := 0 to aList.Items.Count - 1 do begin
    item := aList.Items[i];

    if (aColumnIndex = 0)
    then w := aList.StringWidth(item.Caption)
    else w := aList.StringWidth(item.SubItems[aColumnIndex - 1]);

    if (max_w < w) then max_w := w;
  end;

  if (max_w <> 0)
  then aList.Columns[aColumnIndex].Width := max_w + 16;
end;

{ TSheetList }

constructor TSheetList.Create(AOwner: TComponent; aListMode: TListMode = lmView);
begin
  inherited Create(AOwner);
  Parent := TWinControl(AOwner);

  FListMode := aListMode;

  HandleNeeded;
  Align := alClient;

  if (AOwner is TTabSheet)
  then TTabSheet(AOwner).OnShow := SheetShow;

  FActionAdd := TAction.Create(Self);
  with FActionAdd do begin
    Name := 'actRecAdd';
    Category := 'Tools';
    Caption := 'Add';
    Hint := 'Добавить запись';
    ImageIndex := 3;
    ShortCut := TextToShortCut('Ctrl+I');
    OnExecute := ButtonClick;
  end;

  FActionEdit := TAction.Create(Self);
  with FActionEdit do begin
    Name := 'actRecEdit';
    Category := 'Tools';
    Caption := 'Edit';
    Hint := 'Изменить запись';
    ImageIndex := 4;
    ShortCut := TextToShortCut('Ctrl+Enter');
    OnExecute := ButtonClick;
  end;

  FActionDelete := TAction.Create(Self);
  with FActionDelete do begin
    Name := 'actRecDelete';
    Category := 'Tools';
    Caption := 'Delete';
    Hint := 'Удалить запись';
    ImageIndex := 5;
    ShortCut := TextToShortCut('Ctrl+D');
    OnExecute := ButtonClick;
  end;

  FActionJump := TAction.Create(Self);
  with FActionJump do begin
    Name := 'actLinkJump';
    Category := 'Tools';
    Caption := 'LinkJump';
    Hint := 'Перейти на запись';
    ImageIndex := 28;
    OnExecute := ButtonClick;
  end;

  FActionMoveUp := TAction.Create(Self);
  with FActionMoveUp do begin
    Name := 'actMoveUp';
    Category := 'Tools';
    Caption := 'MoveUp';
    Hint := 'Поместить выше';
    ImageIndex := 29;
    OnExecute := ButtonClick;
  end;

  FActionMoveDown := TAction.Create(Self);
  with FActionMoveDown do begin
    Name := 'actMoveDown';
    Category := 'Tools';
    Caption := 'MoveDown';
    Hint := 'Поместить ниже';
    ImageIndex := 30;
    OnExecute := ButtonClick;
  end;

  //

  FToolBar := TToolBar.Create(Self);
  FToolBar.Parent := Self;
  FToolBar.BorderWidth := 1;
  FToolBar.ButtonHeight := 28;
  FToolBar.ButtonWidth := 30;
  FToolBar.EdgeBorders := [];
  FToolBar.Flat := True;
  FToolBar.Images := fmGEDKeeper.ImageList1;
  FToolBar.ShowHint := True;
  FToolBar.Align := alTop;
  FToolBar.AutoSize := True;

  FBtnMoveDown := TToolButton.Create(Self);
  FBtnMoveDown.Parent := FToolBar;
  FBtnMoveDown.Wrap := True;
  FBtnMoveDown.Action := FActionMoveDown;

  FBtnMoveUp := TToolButton.Create(Self);
  FBtnMoveUp.Parent := FToolBar;
  FBtnMoveUp.Wrap := True;
  FBtnMoveUp.Action := FActionMoveUp;

  FBtnLinkJump := TToolButton.Create(Self);
  FBtnLinkJump.Parent := FToolBar;
  FBtnLinkJump.Wrap := True;
  FBtnLinkJump.Action := FActionJump;

  FBtnDelete := TToolButton.Create(Self);
  FBtnDelete.Parent := FToolBar;
  FBtnDelete.Wrap := True;
  FBtnDelete.Action := FActionDelete;

  FBtnEdit := TToolButton.Create(Self);
  FBtnEdit.Parent := FToolBar;
  FBtnEdit.Wrap := True;
  FBtnEdit.Action := FActionEdit;

  FBtnAdd := TToolButton.Create(Self);
  FBtnAdd.Parent := FToolBar;
  FBtnAdd.Wrap := True;
  FBtnAdd.Action := FActionAdd;

  case FListMode of
    lmView: begin
      FList := TBSListView.Create(Self);
      with TBSListView(FList) do begin
        Parent := Self;
        Align := alClient;
        HideSelection := False;
        ReadOnly := True;
        RowSelect := True;
        SortType := stText;
        ViewStyle := vsReport;
        SortColumn := 0;
        SortDirection := sdAscending;
        ShowSortSign := True;

        OnDblClick := ListDblClick;
        OnKeyDown := ListKeyDown;
      end;
    end;

    lmBox: begin
      FList := TListBox.Create(Self);
      with TListBox(FList) do begin
        Parent := Self;
        Align := alClient;
        Style := lbOwnerDrawVariable;

        OnDrawItem := LBDrawItem;
        OnMeasureItem := LBMeasureItem;

        OnDblClick := ListDblClick;
        OnKeyDown := ListKeyDown;
      end;
    end;
  end;

  FToolBar.Align := alRight;
  FToolBar.List := True;

  SetButtons([lbAdd..lbDelete]);
end;

destructor TSheetList.Destroy;
begin
  FList.Free;

  FBtnDelete.Free;
  FBtnEdit.Free;
  FBtnAdd.Free;

  FToolBar.Free;

  inherited Destroy;
end;

procedure TSheetList.ButtonClick(Sender: TObject);
begin
  if (Sender = FActionAdd) then ItemAdd()
  else
  if (Sender = FActionEdit) then ItemEdit()
  else
  if (Sender = FActionDelete) then ItemDelete()
  else
  if (Sender = FActionJump) then ItemJump()
  else
  if (Sender = FActionMoveUp) then ItemMoveUp()
  else
  if (Sender = FActionMoveDown) then ItemMoveDown();
end;

procedure TSheetList.ListKeyDown(Sender: TObject; var Key: Word; Shift: TShiftState);
begin
  if (ssCtrl in Shift) then
    case Key of
      Ord('I'): ItemAdd();
      Ord('D'): ItemDelete();
      VK_RETURN: ItemEdit();
    end;
end;

function TSheetList.GetSelectedData(): TObject;
begin
  Result := nil;

  case FListMode of
    lmView: begin
      if (TListView(FList).Selected <> nil)
      then Result := TListView(FList).Selected.Data;
    end;
    lmBox: begin
      if (TListBox(FList).ItemIndex > -1)
      then Result := TListBox(FList).Items.Objects[TListBox(FList).ItemIndex];
    end;
  end;
end;

procedure TSheetList.ItemAdd();
begin
  if not(FReadOnly) and Assigned(FOnModify) then FOnModify(Self, nil, raAdd);
end;

procedure TSheetList.ItemEdit();
begin
  if not(FReadOnly) and Assigned(FOnModify) and (GetSelectedData() <> nil)
  then FOnModify(Self, GetSelectedData(), raEdit);
end;

procedure TSheetList.ItemDelete();
begin
  if not(FReadOnly) and Assigned(FOnModify) and (GetSelectedData() <> nil)
  then FOnModify(Self, GetSelectedData(), raDelete);
end;

procedure TSheetList.ItemJump();
begin
  if Assigned(FOnModify) and (GetSelectedData() <> nil)
  then FOnModify(Self, GetSelectedData(), raJump);
end;

procedure TSheetList.ItemMoveDown();
begin
  if not(FReadOnly) and Assigned(FOnModify) and (GetSelectedData() <> nil)
  then FOnModify(Self, GetSelectedData(), raMoveDown);
end;

procedure TSheetList.ItemMoveUp();
begin
  if not(FReadOnly) and Assigned(FOnModify) and (GetSelectedData() <> nil)
  then FOnModify(Self, GetSelectedData(), raMoveUp);
end;

procedure TSheetList.ListDblClick(Sender: TObject);
begin
  ItemEdit();
end;

(*procedure TSheetList.SetActionList(const Value: TActionList);
begin
  {if (Value = nil) then begin
    FActionAdd.ActionList := nil;
    FActionEdit.ActionList := nil;
    FActionDelete.ActionList := nil;
  end else begin
    FActionAdd.ActionList := Value;
    FActionEdit.ActionList := Value;
    FActionDelete.ActionList := Value;
  end;}

  FActionList := Value;
end;*)

procedure TSheetList.SheetShow(Sender: TObject);
begin
  FList.SetFocus();
end;

procedure TSheetList.SetButtons(const Value: TListButtons);
begin
  FButtons := Value;

  FActionAdd.Visible := (lbAdd in FButtons);
  FActionDelete.Visible := (lbDelete in FButtons);
  FActionEdit.Visible := (lbEdit in FButtons);
  FActionJump.Visible := (lbJump in FButtons);
  FActionMoveUp.Visible := (lbMoveUp in FButtons);
  FActionMoveDown.Visible := (lbMoveDown in FButtons);
end;

procedure TSheetList.LBMeasureItem(Control: TWinControl; Index: Integer;
  var Height: Integer);
var
  item: string;
  rt: TRect;
  image: TImage;
  lst: TListBox;
  params: TDrawTextParams;
begin
  // Don't waste time with this on Index = -1
  if (Index > -1) then begin
    lst := TListBox(Control);
    // Create a temporary canvas to calculate the height
    image := TImage.Create(lst);
    try
      rt := lst.ClientRect;
      item := lst.Items.Strings[Index];
      image.Canvas.Font := lst.Font;

      FillChar(params, SizeOf(TDrawTextParams), #0);
      params.cbSize := SizeOf(TDrawTextParams);
      params.iLeftMargin := 3;
      params.iRightMargin := 3;

      Height := DrawTextEx(image.Canvas.Handle, PChar(item), -1, rt, DT_WORDBREAK or DT_CALCRECT, @params);
    finally
      image.Free;
    end;
  end;
end;

procedure TSheetList.LBDrawItem(Control: TWinControl; Index: Integer;
  Rect: TRect; State: TOwnerDrawState);
var
  item: string;
  params: TDrawTextParams;
begin
  TListBox(Control).Canvas.FillRect(Rect);
  item := TListBox(Control).Items.Strings[Index];

  FillChar(params, SizeOf(TDrawTextParams), #0);
  params.cbSize := SizeOf(TDrawTextParams);
  params.iLeftMargin := 3;
  params.iRightMargin := 3;

  DrawTextEx(TListBox(Control).Canvas.Handle, PChar(item), -1, Rect, DT_WORDBREAK, @params);
end;

procedure TSheetList.Columns_BeginUpdate();
begin
  if (FList is TListView)
  then (FList as TListView).Columns.BeginUpdate();
end;

procedure TSheetList.Columns_Clear();
begin
  if (FList is TListView)
  then (FList as TListView).Columns.Clear();
end;

procedure TSheetList.Columns_EndUpdate();
begin
  if (FList is TListView)
  then (FList as TListView).Columns.EndUpdate();
end;

procedure TSheetList.AddColumn(aCaption: string; aWidth: Integer;
  aAutoSize: Boolean);
begin
  if (FList is TBSListView)
  then AddListColumn((FList as TBSListView), aCaption, aWidth, aAutoSize);
end;

procedure TSheetList.SetReadOnly(const Value: Boolean);
begin
  FReadOnly := Value;

  FActionAdd.Enabled := not(FReadOnly);
  FActionDelete.Enabled := not(FReadOnly);
  FActionEdit.Enabled := not(FReadOnly);
  FActionMoveUp.Enabled := not(FReadOnly);
  FActionMoveDown.Enabled := not(FReadOnly);

  if FReadOnly then begin
    if (FList is TListView)
    then (FList as TListView).Color := clBtnFace
    else (FList as TListBox).Color := clBtnFace;
  end else begin
    if (FList is TListView)
    then (FList as TListView).Color := clWindow
    else (FList as TListBox).Color := clWindow;
  end;
end;

end.

File: src/GKUI/Controls/GKSheetList.cs
using System;
using System.Drawing;
using System.Windows.Forms;

using Ext.Utils;
using GKCore;

/// <summary>
/// Localization: clean
/// </summary>

namespace GKUI.Controls
{
	public sealed class GKSheetList : ContainerControl, IDisposable
	{
		public delegate void TModifyEvent(object Sender, object ItemData, TGenEngine.TRecAction Action);

		public enum TListButton : byte
		{
			lbAdd,
			lbEdit,
			lbDelete,
			lbJump,
			lbMoveUp,
			lbMoveDown
		}

		private ToolBarButton FBtnAdd;
		private ToolBarButton FBtnDelete;
		private ToolBarButton FBtnEdit;
		private ToolBarButton FBtnLinkJump;
		private ToolBarButton FBtnMoveUp;
		private ToolBarButton FBtnMoveDown;
		private ToolBar FToolBar;
		private EnumSet FButtons;
		private GKListView FList;
		private GKSheetList.TModifyEvent FOnModify;
		private bool FReadOnly;

		public event GKSheetList.TModifyEvent OnModify
		{
			add
			{
				this.FOnModify = value;
			}
			remove
			{
				if (this.FOnModify == value)
				{
					this.FOnModify = null;
				}
			}
		}

		public EnumSet Buttons
		{
			get { return this.FButtons; }
			set { this.SetButtons(value); }
		}

		public GKListView List
		{
			get { return this.FList; }
		}

		public ToolBar ToolBar
		{
			get { return this.FToolBar; }
		}

		public bool ReadOnly
		{
			get { return this.FReadOnly; }
			set { this.SetReadOnly(value); }
		}

		private void ButtonClick(object sender, ToolBarButtonClickEventArgs e)
		{
			if (object.Equals(e.Button, this.FBtnAdd))
			{
				this.ItemAdd();
			}
			else if (object.Equals(e.Button, this.FBtnEdit))
			{
				this.ItemEdit();
			}
			else if (object.Equals(e.Button, this.FBtnDelete))
			{
				this.ItemDelete();
			}
			else if (object.Equals(e.Button, this.FBtnLinkJump))
			{
				this.ItemJump();
			}
			else if (object.Equals(e.Button, this.FBtnMoveUp))
			{
				this.ItemMoveUp();
			}
			else if (object.Equals(e.Button, this.FBtnMoveDown))
			{
				this.ItemMoveDown();
			}
		}

		private void List_DoubleClick(object sender, EventArgs e)
		{
			this.ItemEdit();
		}

		private void List_KeyDown(object sender, KeyEventArgs e)
		{
			if (e.Control)
			{
				switch (e.KeyCode) {
					case Keys.I:
						this.ItemAdd();
						break;
					case Keys.D:
						this.ItemDelete();
						break;
					case Keys.Return:
						this.ItemEdit();
						break;
				}
			}
		}

		private void SheetShow(object sender, EventArgs e)
		{
			this.FList.Focus();
		}

		private void SetButtons(EnumSet Value)
		{
			this.FButtons = Value;
			this.FBtnAdd.Visible = this.FButtons.InSet(GKSheetList.TListButton.lbAdd);
			this.FBtnDelete.Visible = this.FButtons.InSet(GKSheetList.TListButton.lbDelete);
			this.FBtnEdit.Visible = this.FButtons.InSet(GKSheetList.TListButton.lbEdit);
			this.FBtnLinkJump.Visible = this.FButtons.InSet(GKSheetList.TListButton.lbJump);
			this.FBtnMoveUp.Visible = this.FButtons.InSet(GKSheetList.TListButton.lbMoveUp);
			this.FBtnMoveDown.Visible = this.FButtons.InSet(GKSheetList.TListButton.lbMoveDown);
			this.FToolBar.Visible = !this.FButtons.IsEmpty();
		}

		public object GetSelectedData()
		{
			object result = null;
			if (this.FList.SelectedItem() != null)
			{
				result = this.FList.SelectedItem().Data;
			}
			return result;
		}

		private void SetReadOnly(bool Value)
		{
			this.FReadOnly = Value;
			this.FBtnAdd.Enabled = !this.FReadOnly;
			this.FBtnDelete.Enabled = !this.FReadOnly;
			this.FBtnEdit.Enabled = !this.FReadOnly;
			this.FBtnMoveUp.Enabled = !this.FReadOnly;
			this.FBtnMoveDown.Enabled = !this.FReadOnly;
			if (this.FReadOnly)
			{
				this.FList.BackColor = SystemColors.Control;
			}
			else
			{
				this.FList.BackColor = SystemColors.Window;
			}
		}

		public GKSheetList(Control AOwner)
		{
			AOwner.SuspendLayout();
			this.Dock = DockStyle.Fill;
			AOwner.Controls.Add(this);
			AOwner.ResumeLayout(false);
			base.SuspendLayout();
			this.FBtnMoveDown = new ToolBarButton();
			this.FBtnMoveDown.ImageIndex = 30;
			this.FBtnMoveDown.ToolTipText = LangMan.LSList[300];
			this.FBtnMoveUp = new ToolBarButton();
			this.FBtnMoveUp.ImageIndex = 29;
			this.FBtnMoveUp.ToolTipText = LangMan.LSList[299];
			this.FBtnLinkJump = new ToolBarButton();
			this.FBtnLinkJump.ImageIndex = 28;
			this.FBtnLinkJump.ToolTipText = LangMan.LSList[298];
			this.FBtnDelete = new ToolBarButton();
			this.FBtnDelete.ImageIndex = 5;
			this.FBtnDelete.ToolTipText = LangMan.LSList[21];
			this.FBtnEdit = new ToolBarButton();
			this.FBtnEdit.ImageIndex = 4;
			this.FBtnEdit.ToolTipText = LangMan.LSList[20];
			this.FBtnAdd = new ToolBarButton();
			this.FBtnAdd.ImageIndex = 3;
			this.FBtnAdd.ToolTipText = LangMan.LSList[19];
			this.FToolBar = new ToolBar();
			this.FToolBar.Appearance = ToolBarAppearance.Flat;
			this.FToolBar.Buttons.AddRange(new ToolBarButton[6] {
			                               	this.FBtnAdd,
			                               	this.FBtnEdit,
			                               	this.FBtnDelete,
			                               	this.FBtnLinkJump,
			                               	this.FBtnMoveUp,
			                               	this.FBtnMoveDown});
			this.FToolBar.ImageList = GKUI.TfmGEDKeeper.Instance.ImageList_Buttons;
			this.FToolBar.ShowToolTips = true;
			this.FToolBar.ButtonClick += new ToolBarButtonClickEventHandler(this.ButtonClick);
			this.FList = new GKListView();
			this.FList.Location = new Point(0, 0);
			this.FList.Size = new Size(500, 290);
			this.FList.HideSelection = false;
			this.FList.LabelEdit = false;
			this.FList.FullRowSelect = true;
			this.FList.View = View.Details;
			this.FList.DoubleClick += new EventHandler(this.List_DoubleClick);
			this.FList.KeyDown += new KeyEventHandler(this.List_KeyDown);
			this.FToolBar.Dock = DockStyle.Right;
			this.FList.Dock = DockStyle.Fill;
			base.Controls.Add(this.FList);
			base.Controls.Add(this.FToolBar);
			base.Controls.SetChildIndex(this.FList, 0);
			base.Controls.SetChildIndex(this.FToolBar, 1);
			base.ResumeLayout(false);
			this.SetButtons(EnumSet.Create(new Enum[] { TListButton.lbAdd, TListButton.lbEdit, TListButton.lbDelete }));
		}

		protected override void Dispose(bool Disposing)
		{
			if (Disposing)
			{
				SysUtils.Free(this.FList);
				SysUtils.Free(this.FBtnLinkJump);
				SysUtils.Free(this.FBtnMoveUp);
				SysUtils.Free(this.FBtnMoveDown);
				SysUtils.Free(this.FBtnDelete);
				SysUtils.Free(this.FBtnEdit);
				SysUtils.Free(this.FBtnAdd);
				SysUtils.Free(this.FToolBar);
			}
			base.Dispose(Disposing);
		}

		private void ItemAdd()
		{
			if (!this.FReadOnly && this.FOnModify != null)
			{
				this.FOnModify(this, null, TGenEngine.TRecAction.raAdd);
			}
		}

		private void ItemEdit()
		{
			if (!this.FReadOnly && this.FOnModify != null && this.GetSelectedData() != null)
			{
				this.FOnModify(this, this.GetSelectedData(), TGenEngine.TRecAction.raEdit);
			}
		}

		private void ItemDelete()
		{
			if (!this.FReadOnly && this.FOnModify != null && this.GetSelectedData() != null)
			{
				this.FOnModify(this, this.GetSelectedData(), TGenEngine.TRecAction.raDelete);
			}
		}

		private void ItemJump()
		{
			if (this.FOnModify != null && this.GetSelectedData() != null)
			{
				this.FOnModify(this, this.GetSelectedData(), TGenEngine.TRecAction.raJump);
			}
		}

		private void ItemMoveUp()
		{
			if (!this.FReadOnly && this.FOnModify != null && this.GetSelectedData() != null)
			{
				this.FOnModify(this, this.GetSelectedData(), TGenEngine.TRecAction.raMoveUp);
			}
		}

		private void ItemMoveDown()
		{
			if (!this.FReadOnly && this.FOnModify != null && this.GetSelectedData() != null)
			{
				this.FOnModify(this, this.GetSelectedData(), TGenEngine.TRecAction.raMoveDown);
			}
		}

		public void Columns_BeginUpdate()
		{
		}

		public void Columns_Clear()
		{
			if (this.FList != null)
			{
				this.FList.Columns.Clear();
			}
		}

		public void Columns_EndUpdate()
		{
		}

		public void AddColumn(string aCaption, int aWidth, bool aAutoSize)
		{
			this.FList.AddListColumn(aCaption, aWidth, aAutoSize);
		}
	}
}
